### PR TITLE
Add new ZK opcodes

### PR DESCRIPTION
**Pull request by AurelienFT**
https://github.com/FuelLabs/fuel-specs/issues/615

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog
- [x] New behavior is reflected in tests
- [x] If performance characteristic of an instruction change, update gas costs as well or make a follow-up PR for that
- [x] [The specification](https://github.com/FuelLabs/fuel-specs/) matches the implemented behavior (link update PR if changes are needed)

### Before requesting review
- [x] I have reviewed the code myself
- [x] I have created follow-up issues caused by this PR and linked them here

### After merging, notify other teams

[Add or remove entries as needed]

- [ ] [Rust SDK](https://github.com/FuelLabs/fuels-rs/)
- [ ] [Sway compiler](https://github.com/FuelLabs/sway/)
- [ ] [Platform documentation](https://github.com/FuelLabs/devrel-requests/issues/new?assignees=&labels=new+request&projects=&template=NEW-REQUEST.yml&title=%5BRequest%5D%3A+) (for out-of-organization contributors, the person merging the PR will do this)
- [ ] Someone else?

### CODE REVIEW

File: fuel-asm/src/panic_reason.rs [hunk 1]
```diff
@@ -152,6 +152,11 @@ enum_from! {
         BlobIdAlreadyUploaded = 0x37,
         /// Active gas costs do not define the cost for this instruction.
         GasCostNotDefined = 0x38,
+        /// The curve id is not supported.
+        UnsupportedCurveId = 0x39,
+        // TODO: Maybe add more different errors
+        /// Read alt_bn_128 curve point is invalid.
+        InvalidAltBn128Point = 0x3a,
```
Comment (rymnc) [inline fuel-asm/src/panic_reason.rs:159]: perhaps `InvalidEllipticCurvePoint` since we are going to add support for multiple elliptic curves?

Comment (AurelienFT) [reply fuel-asm/src/panic_reason.rs]: Good idea. edited.

File: fuel-vm/src/interpreter/crypto.rs [hunk 2]
```diff
@@ -202,3 +246,219 @@ pub(crate) fn sha256(
...
+    pc: RegMut<PC>,
+    success: &mut u64,
+    identifier: Word,
+    number_elements: Word,
+    elements_ptr: Word,
+) -> SimpleResult<()> {
+    match identifier {
+        // Optimal ate pairing / alt_bn128
+        0 => {
+            // Each element consistsof an uncompressed G1 point (64 bytes) and an
```
Comment (Dentosal) [inline fuel-vm/src/interpreter/crypto.rs:425]:
```suggestion
            // Each element consists of an uncompressed G1 point (64 bytes) and an
```

File: fuel-vm/src/interpreter/crypto.rs [hunk 3]
```diff
@@ -202,3 +246,219 @@ pub(crate) fn sha256(
...
+                    let point2 = read_g1_point_alt_bn_128(
+                        memory,
+                        points_ptr.checked_add(64).ok_or(
+                            crate::error::PanicOrBug::Panic(
+                                fuel_tx::PanicReason::ArithmeticOverflow,
+                            ),
+                        )?,
+                    )?;
+                    let mut output = [0u8; 64];
+                    #[allow(clippy::arithmetic_side_effects)]
```
Comment (Dentosal) [inline fuel-vm/src/interpreter/crypto.rs:367]: How do we know the addition cannot panic? There should be a `// Safety: ...` comment to explain that.

File: fuel-vm/src/interpreter/crypto.rs [hunk 4]
```diff
@@ -202,3 +246,219 @@ pub(crate) fn sha256(
...
+                        )?,
+                        32u64,
+                    )?)
+                    .map_err(|_| {
+                        crate::error::PanicOrBug::Panic(
+                            fuel_tx::PanicReason::InvalidEllipticCurvePoint,
+                        )
+                    })?;
+                    let mut output = [0u8; 64];
+                    #[allow(clippy::arithmetic_side_effects)]
```
Comment (Dentosal) [inline fuel-vm/src/interpreter/crypto.rs:391]: How do we know the multiplication cannot panic? There should be a `// Safety: ...` comment to explain that.

**Comment by Dentosal** [comment]
I did some polishing in https://github.com/FuelLabs/fuel-vm/pull/874, which could be merged into this one

**Review by Dentosal — CHANGES_REQUESTED** [review]
This is lacking tests for memory boundary violations

File: fuel-vm/src/interpreter/crypto.rs (hunk 4 above)
Comment (AurelienFT) [reply fuel-vm/src/interpreter/crypto.rs:347]: Added a comment.

File: fuel-vm/src/interpreter/crypto.rs (hunk 3 above)
Comment (AurelienFT) [reply fuel-vm/src/interpreter/crypto.rs:324]: Added a comment.

**Comment by AurelienFT** [comment]
@Dentosal I added new tests for memory overflow and writing writes.

**Review by Dentosal — APPROVED** [review]

**Review by rymnc — APPROVED** [review]
i would prefer if we worked with more kats but it can be a follow up pr.

**Pull request merged**